docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,433 @@
+
+### Code formatters enforce formatting conventions
+To both eliminate these kinds of formatting ambiguities and to automate the otherwise-tedious process of formatting code by hand, tools to automatically format code have been developed. These tools take files of source code as input and output the same code, but reformatted as necessary according to a pre-specified set of formatting rules.
+
+In Python, the most popular code formatting tool is called `black`. This tool both automatically formats code and imposes its own opinionated set of formatting rules (hence its tagline: "any color you want"). This eliminates the need for developers to make (or argue about) these decisions themselves. Although `black`'s formatting rules can take some getting used to, the benefit of eliminating almost all formatting-related decisions is usually worth the effort of adjusting to its opinions. 
+
+
+### Setting up `black`
+To install `black` in our conda environment, run:
+```bash
+mamba install -y black
+```
+
+Next, let's add some poorly formatted code to our toy project. Add the following code (from our previous example) to `main.py`:
+```python
+def calculate_total_income(gross_wages, taxable_interest, total_dividends, qualified_dividends, other_income):
+    total_income = gross_wages + taxable_interest + (total_dividends - qualified_dividends) + other_income
+    return total_income
+```
+Now, from the command line, run:
+```bash
+black .
+```
+This command tells `black` to reformat all the Python files in the current directory (the lone dot in the command is the relative path to the current directory). If you look at `main.py` now, you'll see that the code has been reformatted according to `black`'s rules.
+
+Now, try simplifying the calculation of `total_income` by removing the dividends from the sum, then run `black` again. You should see that, because the `total_income` line is shorter, it can fit on one line, so `black` automatically converts it back to a single line.
+
+There is a succinct overview of `black`'s formatting rules in [its documentation](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html). While it's good to be familiar with these rules, there is no need to memorize them. Indeed, the purpose of using a formatting tool like `black` is to eliminate the need to manually write your code according to any particular formatting conventions or even to think about formatting at all. 
```

Comment:
since many people write primarily in notebooks, mention `nb_black`? or if they do it in VScode I guess it's built in?

Reply:
I think it's builtin, and hopefully people are/will be writing notebooks in a repo that was started from our python-project template (which uses ruff, which has builtin support for notebooks)